srcCpp03/RTIDDSImpl.cxx

```diff
@@ -564,6 +589,110 @@ class RTIPublisher: public RTIPublisherBase<T> {
     }
 };
 
+#ifdef RTI_FLATDATA_AVAILABLE
+  template<typename T>
+  class RTIFlatDataPublisher: public RTIPublisherBase<T> {
+  protected:
+      typedef typename rti::flat::flat_type_traits<T>::builder Builder;
+      typedef typename rti::flat::PrimitiveArrayOffset<unsigned char, 4> KeyBuilder;
+      typedef typename rti::flat::PrimitiveSequenceBuilder<unsigned char> BinDataBuilder;
+
+      void add_key(Builder &builder, unsigned long int i) {
+          KeyBuilder key_offset = builder.add_key();
+
+          for (int j = 0; j < KEY_SIZE; ++j) {
+              // The key will be i but splitted in bytes
+              key_offset.set_element(j, (unsigned char) (i >> j * 8));
+          }
+      }
+
+  public:
+      RTIFlatDataPublisher(
+              dds::pub::DataWriter<T> writer,
+              int num_instances,
+              rti::core::Semaphore& pongSemaphore,
+              bool useSemaphore,
+              int instancesToBeWritten,
+              ParameterManager *PM
+              )
+              :
+              RTIPublisherBase<T>(
+                      writer,
+                      num_instances,
+                      pongSemaphore,
+                      useSemaphore,
+                      instancesToBeWritten,
+                      PM)
+      {
+          for (unsigned long int i = 0; i < this->_num_instances; ++i) {
+              Builder builder = rti::flat::build_data(writer);
+              add_key(builder, i);
+
+              T *sample = builder.finish_sample();
+
+              this->_instance_handles.push_back(
+                  this->_writer.register_instance(*sample));
+
+              this->_writer.extensions().discard_loan(*sample);
+          }
+
+          // Register the key of MAX_CFT_VALUE
+          Builder builder = rti::flat::build_data(writer);
+          add_key(builder, MAX_CFT_VALUE);
+          T *sample = builder.finish_sample();
+
+          this->_instance_handles.push_back(
+              this->_writer.register_instance(*sample));
+
+          this->_writer.extensions().discard_loan(*sample);
+      }
+
+      inline bool send(TestMessage &message, bool isCftWildcardKey) {
```

Comment:
Don't use `inline`. `inline` is used for non-member functions whose body is defined in a header. In a member function `inline` is unnecessary and compilers ignore it.

Reply:
Good to know, thank you!